vj4/ui/components/ide/ide.js

```diff
@@ -0,0 +1,448 @@
+      testingData: nextTestingData,
+      currentActiveTabIndex: nextActiveTabIndex,
+      currentTestingInputText: nextTestingData[nextActiveTabIndex].input,
+      currentTestingOutputText: nextTestingData[nextActiveTabIndex].output,
+    });
+  }
+
+  onClickTestCaseTab(idx) {
+    return () => {
+      this.setState({
+        currentActiveTabIndex: idx,
+        currentTestingInputText: this.state.testingData[idx].input,
+        currentTestingOutputText: this.state.testingData[idx].output,
+      });
+    };
+  }
+
+  onUpdateTestingCaseInput(ev) {
+    const newText = ev.target.value;
+    const callback = () => {
+      const nextTestingData = _.cloneDeep(this.state.testingData);
+      nextTestingData[this.state.currentActiveTabIndex].input = newText;
+      this.setState({
+        currentTestingInputText: newText,
+        testingData: nextTestingData,
+      });
+    };
+    if (this.state.testingData.length === 0) {
+      this.onAddTestCase(null, callback.bind(this));
+    } else {
+      callback();
+    }
+  }
+
+  onUpdateTestingCaseOutput(ev) {
+    const newText = ev.target.value;
+    const callback = () => {
+      const nextTestingData = _.cloneDeep(this.state.testingData);
+      nextTestingData[this.state.currentActiveTabIndex].output = newText;
+      this.setState({
+        currentTestingOutputText: newText,
+        testingData: nextTestingData,
+      });
+    };
+    if (this.state.testingData.length === 0) {
+      this.onAddTestCase(null, callback.bind(this));
+    } else {
+      callback();
+    }
+  }
+
+  onTestingSubmit() {
+    if (!this.state.testingData.length) {
+      return;
+    }
+    const dataInput = this.state.testingData.map(el => el.input);
+    const dataOutput = this.state.testingData.map(el => el.output);
+    const pid = window.Context.problemId;
+    post(`/p/${pid}/pretest`, {
+      lang: this.state.language,
+      code: this.state.code,
+      data_input: dataInput,
+      data_output: dataOutput,
+    });
+  }
+
+  onProblemSubmit() {
+    const pid = window.Context.problemId;
+    post(`/p/${pid}/submit`, {
+      lang: this.state.language,
+      code: this.state.code,
+    });
+  }
+
+  toggleDisplayEvaluating() {
+    this.setState({
+      displayEvalutating: !this.state.displayEvalutating,
+    });
+  }
+
+  toggleDisplayTesting() {
+    this.setState({
+      displayTesting: !this.state.displayTesting,
+    });
+  }
+
+  renderTestingTabList() {
+    const ret = [];
+    for (let i = 0; i < this.state.testingData.length; ++i) {
+      ret.push(
+        <div
+          className={`tab ${(this.state.currentActiveTabIndex === i) ? 'tab--active' : ''}`}
+          key={i}
+          onClick={this.onClickTestCaseTab(i).bind(this)}
+        >
+          <span>#{i + 1}</span>
+        </div>
+      );
+    }
+    return ret;
+  }
+
+  renderEvaluatingResultList() {
+    if (this.state.resultData) {
+      return this.state.resultData.map(
+        (rdoc, idx) => <ResultItem
```

Comment:
it may be better to directly pass `rdoc` to the component instead of each property